protocol/contracts/TrucoMatch.sol

```diff
@@ -10,42 +10,132 @@ import "./crypt/DeckCrypt.sol";
 
 contract TrucoMatch  {
 
+    struct player {
+        address playerAddress;
+        uint256 tokensAtStake;
+    }
+    
     struct Match {
-        address[2] players; // player 0 is the creator of the match
+        player[2] players; // player 0 is the creator of the match
         CardsStructs.Deck deck;
         CardsStructs.GameState gameState;
         uint256 tokensAtStake; 
+
     }
 
     IERC3333 trucoEngine;
     DeckCrypt deckCrypt;
-    IERC20 trucoin;
+    IERC20 truCoin;
     Match currentMatch;
 
-    constructor (IERC3333 _trucoEngine, DeckCrypt _deckCrypt, IERC20 _trucoin) {
+    // Events
+    event MatchCreated(address match_address);
+    event MatchStarted(address player1, address player2, uint256 tokensAtStake);
+    event PlayerStaked(address player, uint256 tokensAtStake);
+    event DealStarted(address shuffler);
+    event DealEnded();
+
+
+    constructor (IERC3333 _trucoEngine, DeckCrypt _deckCrypt, IERC20 _truCoin, uint256 _tokensAtStake) {
+
         trucoEngine = _trucoEngine;
         deckCrypt = _deckCrypt;
-        currentMatch.players[0] = msg.sender;
+        truCoin = _truCoin;
+        currentMatch.tokensAtStake = _tokensAtStake;
+        currentMatch.players[0].playerAddress = msg.sender;
+
+        //Mint Sould Bound Token
 
-        //Transfer Trucoins from owner to contract
-        //Mint Sould Bound Token 
+        emit MatchCreated(address(this));
     }
-    
         
-    // Method for joining the match 
+    // Method for second player joining the match 
     function join() public {
-        // Check match state
+        require(currentMatch.players[0].playerAddress != msg.sender, "Match creator is already joined");
+        require(currentMatch.players[1].playerAddress == address(0), "Match is full");
+        require(truCoin.balanceOf(msg.sender) >= currentMatch.tokensAtStake, "You don't have enough Trucoins to join this match");
+        require(truCoin.allowance(msg.sender, address(this)) >= currentMatch.tokensAtStake, "Not enough trucoins transfer approved");
```

Comment:
This two requires checks are duplicated in stake(). I'm thinking of moving this to a modifier or make ir a reusable function that checks preconditions needed for both use cases

Reply:
I think that it would be enough to have this checks just at stake method, I forgot to delete them from join function